test/jdk/java/awt/dnd/DragSourceMotionListenerTest.java

```diff
@@ -0,0 +1,250 @@
+
+    static volatile boolean passedTest1 = false;
+    static volatile boolean passedTest2 = false;
+
+    private static final Point testPoint1 = new Point();
+    private static final Point testPoint2 = new Point();
+    private static volatile Point srcPoint;
+    private static volatile Point dstOutsidePoint;
+    private static volatile Point dstInsidePoint;
+
+    private static final Transferable t = new StringSelection("TEXT");
+    private static final DragGestureListener gestureListener = e -> e.startDrag(null, t);
+
+    private static final DragSourceAdapter sourceAdapter = new DragSourceAdapter() {
+        public void dragMouseMoved(DragSourceDragEvent dsde) {
+            if (Math.abs(dsde.getX() - testPoint1.getX()) < 5) {
+                passedTest1 = true;
+            }
+        }
+
+        public void dragDropEnd(DragSourceDropEvent dsde) {
+            if (Math.abs(dsde.getX() - testPoint2.getX()) < 5) {
+                passedTest2 = true;
+            }
+        }
+    };
+
+    private static final DropTargetListener targetAdapter = new DropTargetAdapter() {
+        public void drop(DropTargetDropEvent e) {
+            e.acceptDrop(DnDConstants.ACTION_COPY);
+            try {
+                final Transferable t = e.getTransferable();
+                final String str =
+                        (String) t.getTransferData(DataFlavor.stringFlavor);
+                e.dropComplete(true);
+            } catch (Exception ex) {
+                ex.printStackTrace();
+                e.dropComplete(false);
+            }
+        }
+    };
+
+    private static final DropTarget dropTarget = new DropTarget(target, targetAdapter);
+    Component clickedComponent = null;
+
+    private void createAndShowUI() {
+        frame = new Frame("DragSourceMotionListenerTest");
+        ds.addDragSourceListener(sourceAdapter);
+        ds.addDragSourceMotionListener(sourceAdapter);
+        ds.createDefaultDragGestureRecognizer(source, DnDConstants.ACTION_COPY, gestureListener);
+        target.setDropTarget(dropTarget);
+
+        frame.setLayout(new GridLayout(1, 2));
+
+        frame.add(source);
+        frame.add(target);
+
+        Toolkit.getDefaultToolkit()
+               .addAWTEventListener(this, AWTEvent.MOUSE_EVENT_MASK);
+        frame.pack();
+        frame.setVisible(true);
+    }
+
+    public static void main(String[] args) throws Exception {
+        final Dimension[] d = new Dimension[1];
+        try {
+            Robot robot = new Robot();
+            robot.setAutoDelay(10);
+
+            DragSourceMotionListenerTest dsmObj = new DragSourceMotionListenerTest();
+            EventQueue.invokeAndWait(dsmObj::createAndShowUI);
+            robot.waitForIdle();
+            robot.delay(1000);
+
+            EventQueue.invokeAndWait(() -> {
+                srcPoint = source.getLocationOnScreen();
+                d[0] = source.getSize();
+                srcPoint.translate(d[0].width / 2, d[0].height / 2);
```

Comment:
The logic hasn't changed. It should rather look like I suggested in [my previous comment](https://github.com/openjdk/jdk/pull/21213#discussion_r1781127928):

```suggestion
                Point p = source.getLocationOnScreen();
                Dimension d = source.getSize();
                p.translate(d.width / 2, d.height / 2);
                srcPoint = p;
```
Otherwise, the `volatile` modifier does not make sense. Prepare the object so that all its field have the correct values, then assign the reference to the object to the volatile field.